### PR TITLE
Model builder for logistic regression & patching sklearn

### PR DESCRIPTION
**Pull request by OnlyDeniko**
Trying to write wrapper for C++ Model builder on Cython

### CODE REVIEW

File: examples/log_reg_model_builder.py [hunk 1]
```diff
@@ -1 +1,94 @@
-# model builders are not available yet, coming soon
\ No newline at end of file
+#*******************************************************************************
+# Copyright 2014-2020 Intel Corporation
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:2]:
```suggestion
# Copyright 2020 Intel Corporation
```

File: examples/log_reg_model_builder.py [hunk 2]
```diff
@@ -1 +1,94 @@
...
+#*******************************************************************************
+
+from sklearn.metrics import * 
+import daal4py as d4p
+import numpy as np
+from sklearn.datasets import load_iris
+from sklearn.datasets import load_breast_cancer
+from sklearn.linear_model import LogisticRegression
+
+def iris_with_intercept():
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:26]: I would suggest taking this out as unit tests. An example should clearly cover one case.

File: examples/log_reg_model_builder.py [hunk 3]
```diff
@@ -1 +1,94 @@
...
+    beta = clf.coef_
+    #print(beta)
+    #print(clf.intercept_)
+    builder = d4p.logistic_regression_model_builder(beta.shape[1], beta.shape[0])
+    builder.setBeta(clf.coef_, clf.intercept_, flag)
+
+    pred = d4p.logistic_regression_prediction(nClasses=2)
+    #print(builder.model)
+    predict_result = pred.compute(X, builder.model)
+    #print(predict_result.prediction)
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:87]:
```python
assert np.allclose(pred_daal, pred_sklearn)
```

File: examples/log_reg_model_builder.py [hunk 4]
```diff
@@ -1 +1,94 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#*******************************************************************************
+
+from sklearn.metrics import *
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:19]: Why?

File: examples/log_reg_model_builder.py [hunk 5]
```diff
@@ -1 +1,94 @@
...
+import daal4py as d4p
+import numpy as np
+from sklearn.datasets import load_iris
+from sklearn.datasets import load_breast_cancer
+from sklearn.linear_model import LogisticRegression
+
+def iris_with_intercept():
+    X, y = load_iris(return_X_y=True)
+    flag = True
+    clf = LogisticRegression(fit_intercept=flag, max_iter=10000, random_state=0).fit(X, y)
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:29]:
```python
n_classes=3
clf = LogisticRegression(fit_intercept=True, random_state=0).fit(X, y)
builder = d4p.logistic_regression_model_builder(n_classes=n_classes, n_features=X.shape[1])
builder.set_beta(clf.coef_, clf.intercept_)

alg_pred = d4p.logistic_regression_prediction(nClasses=n_classes)
pred_daal = alg_pred.compute(X, builder.model).prediction

pred_sklean = clf.predict(X)

assert np.allclose(pred_daal, pred_sklearn)
```

File: examples/log_reg_model_builder.py [hunk 6]
```diff
@@ -1 +1,94 @@
...
+    builder = d4p.logistic_regression_model_builder(beta.shape[1], beta.shape[0])
+    builder.setBeta(clf.coef_, clf.intercept_, flag)
+
+    pred = d4p.logistic_regression_prediction(nClasses=2)
+    #print(builder.model)
+    predict_result = pred.compute(X, builder.model)
+    #print(predict_result.prediction)
+
+
+if __name__ == "__main__":
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:90]: Align the example with others examples.

File: src/modelbuilder.h [hunk 7]
```diff
@@ -20,9 +20,12 @@
 #define _MODELBUILDER_INCLUDED_
 
 #include <daal.h>
+#include <Python.h>
+#include "daal4py.h"
 
 typedef daal::algorithms::gbt::classification::ModelBuilder c_gbt_classification_model_builder;
 typedef daal::algorithms::gbt::regression::ModelBuilder c_gbt_regression_model_builder;
+typedef daal::algorithms::logistic_regression::interface1::ModelBuilder<DAAL_ALGORITHM_FP_TYPE> c_logistic_regression_model_builder;
```
Comment (PetrovKP) [inline src/modelbuilder.h:28]:
```suggestion
typedef daal::algorithms::logistic_regression::ModelBuilder<DAAL_ALGORITHM_FP_TYPE> c_logistic_regression_model_builder;
```

File: src/modelbuilder.h [hunk 8]
```diff
@@ -42,4 +45,16 @@ static daal::algorithms::gbt::regression::ModelPtr * get_gbt_regression_model_bu
     return RAW<daal::algorithms::gbt::regression::ModelPtr>()(obj_->getModel());
 }
 
+template <typename modelFPType = DAAL_ALGORITHM_FP_TYPE>
+static daal::algorithms::logistic_regression::interface1::ModelPtr * get_logistic_regression_model_builder_model(
```
Comment (PetrovKP) [inline src/modelbuilder.h:49]:
```suggestion
static daal::algorithms::logistic_regression::ModelPtr * get_logistic_regression_model_builder_model(
```

File: src/modelbuilder.pyx [hunk 9]
```diff
@@ -43,6 +48,52 @@ cdef extern from "modelbuilder.h":
...
+    cdef c_logistic_regression_model_builder * c_ptr
+    cdef data_management_NumericTablePtr numTableBeta
+
+    def __cinit__(self, size_t nFeatures, size_t nClasses):
+        self.c_ptr = new c_logistic_regression_model_builder(nFeatures, nClasses)
+
+    def __dealloc__(self):
+        del self.c_ptr
+
+    def setBeta(self, numpy.ndarray beta, numpy.ndarray intercept, bool interceptFlag):
```
Comment (PetrovKP) [inline src/modelbuilder.pyx:72]:
```suggestion
    def setBeta(self, beta, intercept):
```

File: src/modelbuilder.pyx [hunk 10]
```diff
@@ -43,6 +48,52 @@ cdef extern from "modelbuilder.h":
...
+        '''
+        Concatenate beta and intercept, convert to daal4py model
+        
+        :param numpy.ndarray beta: beta from scikit-learn model
+        :param numpy.ndarray intercept: intercept from scikit-learn model
+        :param bool interceptFlag: True when scikit-learn model using intercept
+        '''
+        if interceptFlag:
```
Comment (PetrovKP) [inline src/modelbuilder.pyx:80]:
```suggestion
        if np.any(intercept) :
```

File: src/modelbuilder.pyx [hunk 11]
```diff
@@ -19,6 +19,11 @@
 # The model builder object is retrieved through calling model_builder.
 # We will extend this once we know how other model builders willl work in DAAL
 
+import ctypes
```
Comment (PetrovKP) [inline src/modelbuilder.pyx:22]: delete?

File: examples/log_reg_model_builder.py [hunk 12]
```diff
@@ -1 +1,43 @@
...
+    alg_pred = d4p.logistic_regression_prediction(nClasses=n_classes)
+
+    pred_daal = alg_pred.compute(X, builder.model).prediction.flatten()
+    pred_sklearn = clf.predict(X)
+
+    assert np.allclose(pred_daal, pred_sklearn)
+
+
+if __name__ == "__main__":
+    if daal_check_version((2021, 'P', 1)):
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:42]: Check need to move to [run_examples](https://github.com/IntelPython/daal4py/blob/master/examples/run_examples.py)

File: src/modelbuilder.h [hunk 13]
```diff
@@ -20,9 +20,12 @@
 #define _MODELBUILDER_INCLUDED_
 
 #include <daal.h>
+#include <Python.h>
+#include "daal4py.h"
```
Comment (RukhovichIV) [inline src/modelbuilder.h]: do we really need all these imports?

File: generator/gen_daal4py.py [hunk 14]
```diff
@@ -948,18 +948,24 @@ def gen_daal4py(daalroot, outdir, version, warn_all=False, no_dist=False, no_str
...
        'ModelBuilder' in iface.namespace_dict['algorithms::gbt::classification'].classes or \
        'algorithms::gbt::regression' in iface.namespace_dict and \
        'ModelBuilder' in iface.namespace_dict['algorithms::gbt::regression'].classes:
-        with open(jp('src', 'modelbuilder.pyx'), 'r') as f:
-            pyx_modelbuilder = f.read()
+        with open(jp('src', 'gbt_model_builder.pyx'), 'r') as f:
+            pyx_gbt_model_builder = f.read()
+
+    with open(jp('src', 'log_reg_model_builder.pyx'), 'r') as f:
+        pyx_log_reg_model_builder = f.read()
```
Comment (RukhovichIV) [inline generator/gen_daal4py.py:961]: do we need this model builder, when there's no `algorithms::logistic_regression` in our namespace?

File: src/modelbuilder.h (hunk 13 above)
Comment (OnlyDeniko) [reply src/modelbuilder.h]: Removed

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py [hunk 15]
```diff
@@ -959,6 +962,43 @@ def make_ncg_funcs(f, value=False, gradient=False, hessian=False):
...
 
+def daal4py_predict(self, X, resultsToEvaluate):
+    X = check_array(X, accept_sparse='csr', force_all_finite=False)
+
+    try:
+        fptype = getFPType(X)
+    except ValueError:
+        fptype = None
+    
+    if daal_check_version(((2021,'P', 1))) and fptype is not None:
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:973]: check not sparse

File: tests/test_logistic_regression_model_builder.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+        alg_pred = d4p.logistic_regression_prediction(nClasses=n_classes)
+
+        pred_daal = alg_pred.compute(X, builder.model).prediction.flatten()
+        pred_sklearn = clf.predict(X)
+        self.assertTrue(np.allclose(pred_daal, pred_sklearn))
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (PetrovKP) [inline tests/test_logistic_regression_model_builder.py:85]: \n

File: examples/log_reg_model_builder.py [hunk 17]
```diff
@@ -1 +1,55 @@
...
+    # set parameters and train
+    clf = LogisticRegression(fit_intercept=True, max_iter=1000, random_state=0).fit(X, y)
+
+    #set parameters and call model builder
+    builder = d4p.logistic_regression_model_builder(n_classes=n_classes, n_features=X.shape[1])
+    builder.set_beta(clf.coef_, clf.intercept_)
+
+    # set parameters and compute predictions
+    predict_alg = d4p.logistic_regression_prediction(nClasses=n_classes,
+                                                     resultsToEvaluate="computeClassLabels|computeClassProbabilities|computeClassLogProbabilities")
```
Comment (PetrovKP) [inline examples/log_reg_model_builder.py:40]:
```suggestion
                                                     resultsToEvaluate="computeClassLabels")
```

File: daal4py/sklearn/linear_model/_logistic_path_0_22.py [hunk 18]
```diff
@@ -1050,6 +1104,18 @@ def __init__(self, penalty='l2', dual=False, tol=1e-4, C=1.0,
             self.warm_start = warm_start
             self.n_jobs = n_jobs
             self.l1_ratio = l1_ratio
+
+        def predict(self, X):
+            return daal4py_predict(self, X, 'computeClassLabels')
+        
+
+        def predict_log_proba(self, X):
+            return daal4py_predict(self, X, 'computeClassLogProbabilities')
```
Comment (PetrovKP) [inline daal4py/sklearn/linear_model/_logistic_path_0_22.py:1113]: fix proba & log_proba for unpatched case to daal

File: examples/log_reg_model_builder.py (hunk 17 above)
Comment (PetrovKP) [reply examples/log_reg_model_builder.py:40]: Or compare with probabilities from sklearn

File: generator/gen_daal4py.py [hunk 19]
```diff
@@ -942,22 +942,28 @@ def gen_daal4py(daalroot, outdir, version, warn_all=False, no_dist=False, no_str
...
+    pyx_log_reg_model_builder = ''
     if 'algorithms::gbt::classification' in iface.namespace_dict and \
        'ModelBuilder' in iface.namespace_dict['algorithms::gbt::classification'].classes or \
        'algorithms::gbt::regression' in iface.namespace_dict and \
        'ModelBuilder' in iface.namespace_dict['algorithms::gbt::regression'].classes:
-        with open(jp('src', 'modelbuilder.pyx'), 'r') as f:
-            pyx_modelbuilder = f.read()
+        with open(jp('src', 'gbt_model_builder.pyx'), 'r') as f:
+            pyx_gbt_model_builder = f.read()   
+    if iface.version[0] >= 2021 and iface.version[1] >= 2 and \
```
Comment (PetrovKP) [inline generator/gen_daal4py.py]: For gold version. Check please

```suggestion
    if iface.version[0] >= 2021 and iface.version[1] >= 1 and \
```

File: src/log_reg_model_builder.pyx [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+    def set_beta(self, beta, intercept):
+        '''
+        Concatenate beta and intercept, convert to daal4py model
+        
+        :param beta: beta from scikit-learn model
+        :param intercept: intercept from scikit-learn model
+        '''
+        if numpy.any(intercept):
+            tmp = intercept.reshape(-1, 1)
```
Comment (PetrovKP) [inline src/log_reg_model_builder.pyx:54]: `tmp` - bad name
```suggestion
            beta_for_daal= intercept.reshape(-1, 1)
```
For example

File: tests/test_logistic_regression_model_builder.py [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+import numpy as np
+from sklearn.datasets import load_iris
+from sklearn.datasets import load_breast_cancer
+from sklearn.linear_model import LogisticRegression
+from daal4py.sklearn._utils import daal_check_version
+
+from daal4py import _get__daal_link_version__ as dv
+# First item is major version - 2021, second is minor+patch - 0110, third item is status - B
+daal_version = (int(dv()[0:4]), dv()[10:11], int(dv()[4:8]))
+version = ((2021, 'P', 110))
```
Comment (PetrovKP) [inline tests/test_logistic_regression_model_builder.py]: Use it into to code. Because it file will can have other check version

Comment (OnlyDeniko) [reply tests/test_logistic_regression_model_builder.py]: Corrected

File: src/log_reg_model_builder.pyx (hunk 20 above)
Comment (OnlyDeniko) [reply src/log_reg_model_builder.pyx]: Corrected

File: generator/gen_daal4py.py (hunk 19 above)
Comment (OnlyDeniko) [reply generator/gen_daal4py.py]: Corrected

File: examples/log_reg_model_builder.py (hunk 17 above)
Comment (OnlyDeniko) [reply examples/log_reg_model_builder.py]: Corrected

**Review by PetrovKP — APPROVED** [review]

**Review by PetrovKP — APPROVED** [review]

**Pull request merged**